src/main/scala/diplomacy/package.scala

```diff
@@ -7,6 +7,168 @@ import chisel3.Data
 import freechips.rocketchip.config.Parameters
 import scala.language.implicitConversions
 
+
+/**
+  * Diplomacy is a set of abstractions for describing directed, acyclic graphs
+  * where parameters will be negotiated between nodes. These abstractions are
+  * expressed in the form of abstract classes, traits, and type parameters, which
+  * comprises nearly all of the types defined in this file.
+  *
+  * The [[NodeImp]] ("node implementation") is the main abstract type that associates
+  * the type parameters of all other abstract types. Defining a concrete
+  * implementation of [[NodeImp]] will therefore determine concrete types for all
+  * type parameters. For example, passing in a concrete instance of NodeImp to a
+  * SourceNode will fully determine concrete types for all of a SourceNode's type
+  * parameters.
+  *
+  * Specific applications of Diplomacy are expected to either extend these types
+  * or to specify concrete types for the type parameters. This allows for
+  * creating and associating application-specific node, edge, parameter, and bundle types.
+  *
+  *
+  * =Concepts, metaphors, and mnemonics to help with understanding Diplomacy code=
+  *
+  * ==Parameter Types==
+  *  
+  *  There are several types of parameters involved in diplomacy code.
+  *  
+  *  - Up/Down Parameters: These parameter types flow along edges and can be considered as the
+  *    pre-negotiated, unresolved parameters.
+  *  - Edge Paramters: These parameters are the result of the diplomatic negotiation and correspond
+  *    to an edge. These are an abstract concept which carry any sort of conceptual information that is useful to 
+  *    pass along the graph edge. For example, the full address map visible from a given edge and the supported access
+  *    types for each memory region.
+  *  - "p" Parameters: These are parameters of type [[Parameters]] which are implicit and generally available
+  *     in various stages of the diplomacy codebase.
+  *  - Bundle Parameters: These parameters are used for actual hardware generation of the [[Bundle]]s which connect
+  *    diplomatic components. In contrast to edge parameters, this may carry information like the width of an address
+  *    or opcode field.
+  *  
+  * ==Inward/Outward vs. Upward/Downward==
+  *
+  * Diplomacy defines two dimensions: inward/outward and upward/downward.
+  *
+  * Inward/outward refer to the direction of edges from the perspective of a given node.
+  * For a given node:
+  * - Inward refers to edges that point into a node.
+  * - Outward refers to edges that point out of a node.
+  *  
+  *  Therefore, a given edge is inward for one node and outward to another.
+  *
+  * Upward/downward refer to the direction of the overall directed acyclic graph.
```

Comment:
add `Because each each edge is directed, we say that the direction is a flow from sources (nodes that only have outward edges) to sink (nods that only have inward edges).